Add draft of 'how we built batch changes' (#2884)
sourcegraph/about: blogposts/2021/how-we-built-batch-changes.md

Instead, working iteration-based, you define roughly where you want to go and
then take a single small step. Then you look around, get your bearings, correct
course if necessary, and take another step. Build something, ship it, learn,
build again.

I personally love working like this and think that in a lot of contexts in which
software is being developed (fast-moving startups, for example) it's the best way to
ship successful products.

But there's one huge problem with this approach: it's really easy to iterate
yourself into a corner. If all you do is take small steps you can miss the point
where you should take big leaps to someplace else entirely. You end up with [a
local maximum](https://elezea.com/2012/09/iphone-5-local-maximum/). On top of
that: even if you realise that you're stuck and _know_ that you must take a big
leap, it's hard to abandon the things you've built in many, many small
iterations. "So much work went into this! We can't simply delete it! Right?"
Right, it's the [sunken cost
fallacy](https://en.wikipedia.org/wiki/Sunk_cost#Fallacy_effect).

Back to the five pull requests and how we've managed to defy both of these dangers.

> In September 2019 we started working on Batch Changes, except that back then it

Reviewer: ```suggestion
In September 2019 after Eric Brody-Moore had the original idea we started working on Batch Changes, except that back then it
```

Author: ```
In September 2019, _despite_ Eric Brody-Moore, we started working on Batch Changes, except that back then it
```

Reviewer: Hmm it seems like you're trying to re-write history here 🤔

Author: That's _exactly_ what I'm trying to do, yes!

---
title: 'How we built Batch Changes, explained in 5 pull requests'
author: Thorsten Ball
authorUrl: https://twitter.com/thorstenball
publishDate: 2021-04-16T18:00+02:00
tags: [blog]
slug: how-we-built-batch-changes
# TODO: heroImage: /blog/TODO.jpg
# TODO: socialImage: https://about.sourcegraph.com/blog/TODO.jpg
published: false
---

Straightforward is the last word I'd use to describe how we built Batch Changes.
Exciting, rewarding, surprising, iterative, customer-focused - one of those
maybe, but not straightfoward. The two name changes in 1.5 years of development
— from Automation to Campaigns to Batch Changes — give you a glimpse of why. The

> real story, though, the one that makes me doubt whether we actually pulled this

Author: Good point! I rephrased this.

_Sprint_, _Shape Up_ or something else, the idea is this: until you've built the
product, you don't _really_ know what it should look like or how to best build
it. You've never used it — how would you know how it works best? You also
haven't built it yet, which means you don't have experience building it. Chances
are low that you can accurately predict how its development will go.

So you do is you acknowledge that you don't know. And then you lean into that:
you _don't_ plan the development of the product from start to finish in every
detail. You _don't_ nail down and define the hundreds of steps it will take to
go from zero to fully-built and shipped and launched product. Instead, working
in iterations, you define roughly where you want to go and then take a single
small step. Then you look around, get your bearings, correct course if
necessary, take another step. Build something, ship it, learn, build again.

I personally love working like this and think that in a lot of contexts in which
software is being developed (fast-moving startups, for example) it's the best
way to ship successful products.

But there's one huge problem with this approach: it's really easy to iterate

> yourself into a corner. If all you do is take small steps, you can miss the

Author: This kind of feedback is _my jam_ as they say (or ... said? I don't know)

I changed this paragraph to make the metaphor more consistent, please take a look.

(I also had more ideas and actually wanted to bring "dance" into this — "While you're happily doing this two-steps-forward-two-steps-back dance, you might miss when ..." — but let's see if this works too. I'm _always_ happy to go all in on metaphors.)

> yourself into a corner. If all you do is take small steps, you can miss the

Reviewer 2: I left one small suggestion, but otherwise it looks good!

[sunken cost fallacy](https://en.wikipedia.org/wiki/Sunk_cost#Fallacy_effect).

Back to the five pull requests and how we've managed to defy both of these dangers.

In September 2019 we started working on Batch Changes, except that back then it
was called _Automation_ and it was already implemented, kinda: [Quinn][quinn]
had built an impressive prototype to show to customers and ask them whether
they'd use something like this. More than one said "yes, I would, and _exactly
like that_". "[Build one to throw
away](https://wiki.c2.com/?PlanToThrowOneAway)" is what they say - but do you?
Do you really throw away a prototype that's more than 10k lines of code and that
customers pointed at and said "I want this"?

We did, with [this pull request][pr1] — a placeholder here for all of the code
we wrote in September and October 2019 to rebuild the functionality of the
prototype from scratch. What made us do it though? We came to the conclusion
that building the product from the ground up, fully understanding and owning the
code is more important to the long-term success of the project and the team than
merging a prototype to get something out there as fast as possible. From my

> perspective now this was one of the key decisions that, to this day, helped us

Author: Thanks!

avoid accumulating large amounts of technical debt, even though we were
constantly shipping new things.

Only a few months later, though, we were stuck. We realised that what we had
built and shipped as an alpha to customers just didn't work on a conceptual
level. In that version, campaigns (now called batch changes) were executed on
the Sourcegraph instance to produce changes in repositories, which was cool from
a technical standpoint but it felt clunky, and slow, and adding a new type of

> campaign (to produce a new type of change) required us to add new code. "Meh" is

Author: Rephrased the sentence to avoid the name.

customers that were most interested in making large-scale changes often had
custom tooling already. With the CLI producing diffs by running any Docker
container we could give them a place to plug it in — making it large-scale.

This radical change to how the product worked, was, in my opinion, maybe the
most important one made in the life of Batch Changes. The mental model and the
abstraction layers it introduced turned out to be incredibly well-fitting. They
allowed us to go faster and get a _valuable_ tool into the hands of our
customers even earlier.

So far, so good, right? Yes, except, you know, [naming is
hard](https://martinfowler.com/bliki/TwoHardThings.html) and while we could've
made the old name for the feature — _Automation_ — work we decided to rename it.
Take a look at [PR #3][p3] for the first of many in February 2020 that changed
"[Aa]utomation" to "[Cc]ampaigns" in our codebase.

What followed were months of fine-tuning, user testing, shipping, tweaking -
_iterating_. We added more features, fixed a lot of bugs, made things go faster.

But — and by now you should know where this is going — something felt _off_.

> While Campaigns worked and we had customers using it and saying they loved it,

Reviewer 2: Just wanted to say that, in this case and the next several that follow, I like that you are using "Campaigns", because you are talking about renaming. So I wouldn't change it in these places. 🙂

admit that they were confused with the workflow. Some of the questions we got:
why JSON (yes, we made users write JSON by hand)? Why can't I put everything in
that JSON file, why do I have to switch to the web UI? And what if I want to
update the changes?

I wouldn't say we were stuck again - we were constantly shipping new features
and improvements. But it also started dawning on us that maybe we've exhausted
one idea. Tweaking wouldn't get us out of this, nor a sledgehammer. But maybe if
we look at it from a different angle? What if instead of requiring users say
what should happen to produce a campaign, we switch to a _declarative_ model and
let them describe what the campaign should like in a YAML file?

[This pull request](https://github.com/sourcegraph/sourcegraph/pull/10921) shows
how we asked "what if?": we wrote the documentation first (yes, README-driven
development) and showed it to teammates and colleagues. The time it took a
colleague to go from "ok, tell me what campaigns are" to "ohhh, I get it! nice"
was reduced tenfold. Very much to our relief, since the changes necessary to
make the documentation a reality were huge. We had to build a distributed,

> _declarative_ system that manages hundreds and thousands of pull and merge

Reviewer 2: ```suggestion
_declarative_ system that manages hundreds or thousands of pull and merge
```
"and" isn't wrong here, but for some reason "or" sounds more idiomatic to me

requests across different code hosts.

> Once the idea was validated by a lot "this totally makes sense" from potential

Reviewer 2: ```suggestion
Once the idea was validated by a lot of "this totally makes sense" from potential
```

> users we got to work. Look at [this PR][pr4b], number 4b in this series if you

Reviewer 2: ```suggestion
users, we got to work. Look at [this PR][pr4b], number 4b in this series if you
```

will, to see an example of what "declarative" means and how we reconcile the
current and the desired state (as described by a user) of a changeset. (Sidenote
for the curious: the code today is [even cooler][evencoolercode], take a look).

That brings us to the last of the five pull requests.

Towards the end of last year we ripped the "beta" label off of Campaigns and
started to concentrate on getting more customers to use it: writing better (or

> even any) documentation, improving the onboarding process, providing

Reviewer 2: ```suggestion
in some cases any) documentation, improving the onboarding process, providing
```

> troubleshooting help, fixing bugs and edge cases.

Reviewer 2: ```suggestion
troubleshooting help, and fixing bugs and edge cases.
```

> At the same time our fast-growing product marketing team started working towards

Reviewer 2: ```suggestion
At the same time, our fast-growing product marketing team started working towards
```
This comma means there's only one way to parse this. Otherwise I have to read ahead to know how to parse it. Because maybe you were saying, "At the same time *that* our fast-growing..." in which case the sentence ends quite abruptly because the second half of it is missing.

making our product as a whole more consistent. One thing that stood out was the

> name "Campaigns". Next to the names of our other features it didn't fit in. It

Reviewer 2: ```suggestion
name "Campaigns". Next to the names of our other features, it didn't fit in. It
```

> also always required a sentence or two of explanation ("Glad you asked. A

Reviewer 2: ```suggestion
also always required a sentence or two of explanation. ("Glad you asked. A
```

> campaign is a ..."). In a data-driven approach to finding a new name Batch

Reviewer 2: ```suggestion
campaign is a ...") In a data-driven approach to finding a new name, Batch
```

Changes came out as the winner. Ahead of Campaigns, Batches, Clusters, Bulk

> Changes, and others. No one asked in a survey needed an explanation and nearly

Reviewer 2: ```suggestion
Changes, and others. No one we asked in our survey needed an explanation, and nearly
```

> everyone understood what the feature is roughly about.

Reviewer 2: ```suggestion
everyone understood what the feature was roughly about.
```

But renaming what we built _again_? It would've been tempting to answer with
"customers are already using it, why bother?" or "we've recorded demo videos
with the old names!" or "there's so many screenshots we would need to change".

> In all honesty, though, we had to admit: existing customers probably won't mind

Reviewer 2: ```suggestion
In all honesty, though, we had to admit: existing customers probably wouldn't mind
```

as long as it's not a breaking change, some of our screenshots were outdated
already, and we've been meaning to record an up-to-date demo video anyway.

So, [renaming we did][pr5] and followed it up with the first official,
non-alpha, non-beta, download-it-now-and-try-it launch of [Batch
Changes][launch].

> Looking back over these five pull requests and at the past 1.5 years now I'm

Reviewer 2: ```suggestion
Looking back over these five pull requests and at the past 1.5 years now, I'm
```

still surprised. I know how easy it is to say "but this is what we said we

> wanted to build" or "but we invested all that time!". I've worked in iterations

Reviewer 2: ```suggestion
wanted to build" or "but we invested all that time!" I've worked in iterations
```

all my life. 2 week sprints, 4 weeks, Agile, Scrum, even - ugh - [SAFe][safe].

> The danger to get stuck and not knowing how to get out of it is always there.

Reviewer 2: ```suggestion
The danger of getting stuck and not knowing how to get out of it is always there.
```

I'm still not quite sure how we avoided that trap. My best guess (and this
_will_ sound cheesy) is: commitment. Commitment to building an excellent
product, commitment to building something that provides value to customers, even

> if it means starting from scratch when you realise you're in a dead end.

Reviewer 2: ```suggestion
if it means starting from scratch when you realize you're at a dead end.
```

product, you don't _really_ know what it should look like or how to best build
it. You've never used it — how would you know how it works best? You also
haven't built it yet, which means you don't have experience building it. Chances
are low that you can accurately predict how its development will go.

So what you do is you acknowledge that you don't know. And then you lean into that:
you _don't_ plan the development of the product from start to finish in every
detail. You _don't_ nail down and define the hundreds of steps it will take to
go from zero to fully-built and shipped and launched product. Instead, working
in iterations, you define roughly where you want to go and then take a single
small step. Then you look around, get your bearings, correct course if
necessary, take another step. Build something, ship it, learn, build again.

I personally love working like this and think that in a lot of contexts in which
software is being developed (fast-moving startups, for example) it's the best
way to ship successful products.

But there's one huge problem with this approach: if you're constantly taking
small steps and adjusting the direction slightly, you _feel_ productive even

> when you're going in circles. You're constantly doing two steps forward and two

Reviewer 2: ```suggestion
when you're going in circles. You keep taking two steps forward and two
```

---

> title: 'How we built Batch Changes, explained in 5 pull requests'

Reviewer 3: or just "How we avoided the biggest dangers of iteration-base product development"

> title: 'How we built Batch Changes, explained in 5 pull requests'

Author: I really think we should keep the "5 pull requests" in there somehow, because that's what would make me click. "iteration-based product development" in a header is something I would skip. But "you build something in 5 PRs? which ones?" is what would make me click & read.

> title: 'How we built Batch Changes, explained in 5 pull requests'

Reviewer 2: Agreed, I was immediately curious: a story told in 5 PRs? Which ones?

> title: 'How we built Batch Changes, explained in 5 pull requests'

Reviewer 3: What about: 

"How we avoided the biggest dangers of iteration-based product development explained in 5 pull requests"

> title: 'How we built Batch Changes, explained in 5 pull requests'

Reviewer 3: Shorter version: "Avoiding the pitfalls of iteration-based development explained in 5 pull requests"

> title: 'How we built Batch Changes, explained in 5 pull requests'

Author: Hm, I think it depends on the audience we're targeting here. "How we built X" is always interesting to me, especially on a tech company blog. Stuff that sounds "product development-y" is hit and miss to me.

My original idea was:

```
How we built Batch Changes - in 5 pull requests
```

> title: 'How we built Batch Changes, explained in 5 pull requests'

Reviewer 3: My concern with leading with Batch Changes is that many people have never heard of it. Our audience for this post is software engineers, but not necessarily currents users, or even someone who has heard of Sourcegraph before. 

My question for you: is "how we built X" is universally interesting, regardless of product and company familiarity?

> title: 'How we built Batch Changes, explained in 5 pull requests'

Author: > My question for you: is "how we built X" is universally interesting, regardless of product and company familiarity?

I wouldn't say regardless of company familiarity. When I know that the company builds interesting stuff I will definitely be more likely to read.

But given two options:

1. "How we built X" with X being a thing that sounds like a product name I don't know about
2. more generic sounding title

and the same company familiarity, I'd still prefer 1 more.

> title: 'How we built Batch Changes, explained in 5 pull requests'

Reviewer 4: I love the "in 5 pull requests" idea. This immediately hooks me to the article. Wait. What? Oh, nice. What I like from a more analytical perspective is that this title doesn't sound like other titles I'm used too. And I want to know why!! So I keep on reading.

> description: A 300 character limit field for describing your post. Use this is you want to specially craft the excerpt shown on the index page. Uses the first 300 characters of text from your post if this field does not exist.

Author: Done!

author: Thorsten Ball
authorUrl: https://twitter.com/thorstenball
publishDate: 2021-04-16T18:00+02:00
tags: [blog]
slug: how-we-built-batch-changes
# TODO: heroImage: /blog/TODO.jpg
# TODO: socialImage: https://about.sourcegraph.com/blog/TODO.jpg
published: false
---

> "Straightforward" is the last word I'd use to describe how we built Batch

Author: Done!

show how we avoided two big dangers of iteration-based product development:
local maximums and the sunken cost fallacy.

To give you some context for these terms, let me tell you how I understand the
premise of iteration-based work. It doesn't matter whether you call it _Agile_,
_Sprint_, _Shape Up_ or something else, the idea is this: until you've built the
product, you don't _really_ know what it should look like or how to best build
it. You've never used it — how would you know how it works best? You also
haven't built it yet, which means you don't have experience building it. Chances
are low that you can accurately predict how its development will go.

So what you do is you acknowledge that you don't know. And then you lean into that:
you _don't_ plan the development of the product from start to finish in every
detail. You _don't_ nail down and define the hundreds of steps it will take to
go from zero to fully-built and shipped and launched product. Instead, working
in iterations, you define roughly where you want to go and then take a single
small step. Then you look around, get your bearings, correct course if

> necessary, take another step. Build something, ship it, learn, build again.

Author: > Speaking as the author of the post you don't need to add qualifiers.

I still would like to leave them in, since I don't think it's a universal explanation for the concept (others _have_ different definitions) but a rather subjective view.

> necessary, take another step. Build something, ship it, learn, build again.

Author: I moved the paragraphs up, incorporated some wording suggestions, but left others in. Let me know if that works.

> I personally love working like this and think that in a lot of contexts in which

Author: Done!

software is being developed (fast-moving startups, for example) it's the best
way to ship successful products.

But there's one huge problem with this approach: if you're constantly taking
small steps and adjusting the direction slightly, you _feel_ productive even
when you're going in circles. You're constantly doing two steps forward and two

> steps back but never take a big leap to someplace else entirely. You end up with

Author: Done!

> [a local maximum][localmaximum]. And even if you realise that you're stuck and

Author: I added a short explanation, but kept the link. Let me know if it makes sense 😬

_know_ that you must take a big leap, it's hard to abandon the things you've
built in many, many small iterations. "So much work went into this! We can't

> simply delete it! Right?" Right, says the [sunken cost fallacy][sunkencost].

Author: Added a sentence that I hope is clear - let me know if that's not the case

> simply delete it! Right?" Right, says the [sunken cost fallacy][sunkencost].

Author: I kinda rephrased this a bit before I saw this exact comment, but I still think it works in the new structure - let me know if that's not the case.

> 

Author: Done!

Back to the five pull requests and how we've managed to defy both of these dangers.

> 

Reviewer 3: It would be nice to capture screen shots of the pull requests if they can help the reader better understand the story.

> 

Author: Good idea! I uploaded screenshots of the PRs and added them at the top of each section.

[We threw half of it away][pr2]. The trigger was this idea: what if we simply
don't produce changes on the server-side but instead accept _diffs_ through
the API and, as part of the Sourcegraph CLI, provide a tool to produce those
diffs wherever you can run the CLI with whatever Docker container you want?

The server-side would then only have concern itself with diffs: how to
publish them on code hosts as pull and merge requests, how to sync them, how to
track them, how to rate limit their creation, etc. We wouldn't have to ship a
new release every time we came up with a new way to produce diffs. And our
customers that were most interested in making large-scale changes often had
custom tooling already. With the CLI producing diffs by running any Docker
container we could give them a place to plug it in — making it large-scale.

This radical change to how the product worked, was, in my opinion, maybe the
most important one made in the life of Batch Changes. The mental model and the
abstraction layers it introduced turned out to be incredibly well-fitting. They
allowed us to go faster and get a _valuable_ tool into the hands of our
customers even earlier.

> 

Reviewer 3: ```suggestion
## Pull requests 3 & 4: Still not there... more fine-tuning and user testing
```

Author: I used two headings for #3 and #4 (and also realised that PR#4 had the wrong URL)

what should happen to produce a campaign, we switch to a _declarative_ model and
let them describe what the campaign should like in a YAML file?

[This pull request](https://github.com/sourcegraph/sourcegraph/pull/10921) shows
how we asked "what if?": we wrote the documentation first (yes, README-driven
development) and showed it to teammates and colleagues. The time it took a
colleague to go from "ok, tell me what campaigns are" to "ohhh, I get it! nice"
was reduced tenfold. Very much to our relief, since the changes necessary to
make the documentation a reality were huge. We had to build a distributed,
_declarative_ system that manages hundreds or thousands of pull and merge
requests across different code hosts.

Once the idea was validated by a lot of "this totally makes sense" from potential
users, we got to work. Look at [this PR][pr4b], number 4b in this series if you
will, to see an example of what "declarative" means and how we reconcile the
current and the desired state (as described by a user) of a changeset. (Sidenote
for the curious: the code today is [even cooler][evencoolercode], take a look).

That brings us to the last of the five pull requests.

> 

Reviewer 3: ```suggestion
## Pull request 5: Main point of this PR reference
```

Author: I have to admit that it took me really long to understand that "Main point of this PR reference" was directed at me 😄

Exciting, rewarding, surprising, iterative, customer-focused - one of those
maybe, but not straightfoward.

Why? The two name changes in 1.5 years of development - from Automation to
Campaigns to Batch Changes - give you a glimpse. The real story, though, the one
that makes me doubt whether we actually pulled this off, I think is best told by
five pull requests.

Not because these pull requests solve hard technical problems in two lines of
clever code or fix nasty bugs that held us back for months. It's because they
show how we avoided two big dangers of iteration-based product development:
local maximums and the sunken cost fallacy.

To give you some context for these terms, let me tell you how I see the premise
of iteration-based product development. It doesn't matter whether you call it
_Agile_, _Sprint, _Shape Up_ or something else, the idea is this: until you've
built the product, you don't _really_ know what it should look like or how to
best build it. You've never used it, so how would you know how it best works?
You also haven't built it (yet), so you don't have experience building it.
Chances that you can accurately predict how its development will go are low.

> Acknowledging that you _don't_ plan the development of the product from start to

Author: Ah, I see what you mean. Rephrased it and (hopefully) improved it.

In September 2019 we started working on Batch Changes, except that back then it
was called _Automation_ and it was already implemented, kinda: [sqs](TODO) had
built an impressive prototype to show to customers and ask them whether they'd
use something like this. More than one said "yes, I would, and _exactly like
that_". "[Build one to throw away](https://wiki.c2.com/?PlanToThrowOneAway)" is
what they say - but do you? Do you really throw away a prototype that's more
than 10k lines of code and that customers pointed at and said "I want this"?

We did, with [this pull request][pr1], which is a placeholder for all of the
code we wrote in September and October 2019 to rebuild the functionality of the
prototype from scratch. Why did we do it? We thought that building the product
from the ground up, fully understanding and owning the code is more important to
the long-term success of the project and the team than merging a prototype to
get something out there as fast as possible.

Only a few months later, though, we were stuck. We realised that what we had
built and shipped as an alpha to customers just didn't work on a conceptual

> level. In that version, campaigns (now called batch changes) were executed on

Reviewer: The next question that could be answered here is `What went "wrong" (though it's not really wrong) in the product development that you built something that didn't work on a conceptual level?`

Author: Ah, yes. That _is_ interesting. I'm not sure I have an answer though (or if someone else could better write that), but I'll think about it.

Author: I can't up with something that I can confidently write about. I'll keep thinking about it though.

_declarative_ system that manages hundreds and thousands of pull and merge
requests across different code hosts.

Once the idea was validated by a lot "this totally makes sense" from potential
users we got to work. Look at [this PR][pr4b], number 4b in this series if you
will, to see an example of what "declarative" means and how we reconcile the
current and the desired state (as described by a user) of a changeset. (Sidenote
for the curious: the code today is [even cooler][evencoolercode], take a look).

That brings us to the last of the five pull requests. 

Towards the end of last year we ripped the "beta" label off of Campaigns and
started to concentrate on getting more customers to use it: writing better (or
even any) documentation, improving the onboarding process, providing
troubleshooting help, fixing bugs and edge cases.

At the same time our fast growing product marketing team started working towards
making our product as a whole more consistent. One thing that stood out was the
name "Campaigns" that didn't fit next to the names of our other features and
always required a sentence or two of explanation ("Glad you asked. A campaign is

> a ..."). The name Batch Changes, on the other hand, fared far better in surveys:

Reviewer: Maybe this isn't the time or place but could be cool to throw in a few sentences about the naming process. Or at least a couple other options we were considering!

Author: "Clusters" and "Bulk Changes" and "Batches" and "Campaigns" were other options and didn't do as well. I'll try to work that in!

A separate "Naming is hard, or: how to find a product name" blog post would be cool.

Author: Changed that paragraph

still inspires and motivates me, because I'd rather build and ship something
that _is_ valuable and that _is_ good than to give customers something just
because we build it.

[pr1]: https://github.com/sourcegraph/sourcegraph/pull/5482
[pr2]: https://github.com/sourcegraph/sourcegraph/pull/8008
[pr3]: https://github.com/sourcegraph/about/pull/583
[pr4]: https://github.com/sourcegraph/sourcegraph/pull/11972
[pr4b]: https://github.com/sourcegraph/sourcegraph/pull/11972
[evencoolercode]: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@e863448757e09850349b8a2bd7b1e540f6a6259a/-/blob/enterprise/internal/batches/reconciler/executor.go#L91-129
[pr5]: https://github.com/sourcegraph/about/pull/2745
[launch]: https://about.sourcegraph.com/blog/introducing-batch-changes/
[safe]: https://www.google.com/search?q=scaled+agile+framework&tbm=isch

<!---
Pull requests:

PR #1:
- 5 Sep 2019: [a8n: Implement campaigns in GraphQL API by tsenart](https://github.com/sourcegraph/sourcegraph/pull/5482)
PR #2:
- 26 Jan 2021: [a8n: support creating campaign plans from API only, not UI by sqs](https://github.com/sourcegraph/sourcegraph/pull/8008)
PR #3:

> - 20 Feb 2021: [rename (in docs and settings): automation -> campaigns by sqs](https://github.com/sourcegraph/sourcegraph/pull/8507)

Reviewer 5: these two years should probably be 2020?

Author: Yes! But these are just comments/notes that I used when collecting relevant PRs.
